File: frontend/src/types/supplierProduct.ts
import type { ProductUnit } from "./product";

export interface SupplierProduct {
  idSupplierProduct: number;
  idProduct: number;
  idSupplier: number;
  supplierReference: string;
  unitPrice: number;
  minimumOrderQuantity: number;
  expectedLeadTimeDays: number;
  packagingQuantity: number;
  packagingUnit: ProductUnit;
  fractionable: boolean;
  active: boolean;
}

export interface SupplierProductCreate {
  idProduct: number;
  idSupplier: number;
  supplierReference?: string;
  unitPrice?: number;
  minimumOrderQuantity?: number;
  expectedLeadTimeDays?: number;
  packagingQuantity?: number;
  packagingUnit?: ProductUnit;
  fractionable: boolean;
}

export interface SupplierProductUpdate {
  idProduct: number;
  idSupplier: number;
  supplierReference?: string;
  unitPrice?: number;
  minimumOrderQuantity?: number;
  expectedLeadTimeDays?: number;
  packagingQuantity?: number;
  packagingUnit?: ProductUnit;
  fractionable: boolean;
  active: boolean;
}

File: backend/buylogic/src/main/java/com/buylogic/model/SupplierProduct.java
package com.buylogic.model;

import java.math.BigDecimal;
import java.time.LocalDateTime;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Table(
    name = "supplier_product",
    uniqueConstraints = {
        @UniqueConstraint(
            name = "uq_supplier_product",
            columnNames = {"id_product", "id_supplier"}
        )
    }
)
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class SupplierProduct {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer idSupplierProduct;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "id_product", nullable = false)
    private Product product;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "id_supplier", nullable = false)
    private Supplier supplier;

    @Column(name = "supplier_reference", length = 100)
    private String supplierReference;

    @Column(name = "unit_price", precision = 12, scale = 2)
    private BigDecimal unitPrice;

    @Column(name = "minimum_order_quantity", precision = 14, scale = 3)
    private BigDecimal minimumOrderQuantity;

    @Column(name = "expected_lead_time_days")
    private Integer expectedLeadTimeDays;

    @Column(nullable = false)
    private Boolean active = true;

    @Column(nullable = false)
    private LocalDateTime createdAt;

    @Column(nullable = false)
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        LocalDateTime now = LocalDateTime.now();
        createdAt = now;
        updatedAt = now;
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}
